Add "Anatomy of a Service" documentation (#1782)
awslabs/smithy-rs: design/src/docs/anatomy.md

# The Anatomy of a Service

What is [Smithy](https://awslabs.github.io/smithy/2.0/index.html)? At a high-level, it's a grammar for specifying services modulo business logic. A [Smithy Service](https://awslabs.github.io/smithy/2.0/spec/service-types.html#service) specifies a set of function signatures in the form of [Operations](https://awslabs.github.io/smithy/2.0/spec/service-types.html#operation) to which encapsulate business logic. A Smithy implementation should, for each Smithy Service, provide a builder, which accepts functions conforming to said signatures, and returns a service subject to the semantics specified by the model.

> This is an dissection of the various mechanisms at work inside Smithy Rust in order to provide such a builder. This is intended for new contributors or users who are interested in the internal details.

Author: This should be changed to emphasize that this is not a description of the logic/implementation details of the Kotlin code but rather a description of how the Rust composes together and it's relation to the Smithy model.

    fn from_request(request: http::Request) -> Self::Future;
}

/// A protocol aware function taking `self` to [`http::Response`].
pub trait IntoResponse<Protocol> {
    /// Performs a conversion into a [`http::Response`].
    fn into_response(self) -> http::Response<BoxBody>;
}
```

Note that both traits are parameterized by `Protocol`. These are ZST marker structs, a few [publicly available](https://awslabs.github.io/smithy/2.0/aws/protocols/index.html) markers are:

```rust
/// [AWS REST JSON 1.0 Protocol](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-restjson1-protocol.html).
pub struct AwsRestJson1;

/// [AWS REST XML Protocol](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-restxml-protocol.html).
pub struct AwsRestXml;

/// [AWS JSON 1.0 Protocol](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-json-1_0-protocol.html).
pub struct AwsJson10;

/// [AWS JSON 1.1 Protocol](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-json-1_1-protocol.html).
pub struct AwsJson11;
```

## Upgrading a Model Service

Notice that we can "upgrade" a model service to a HTTP service using `FromRequest` and `IntoResponse` described in the prior section:

> ![Upgrade Data Flow Diagram](imgs/upgrade-dfd.png)

Author: There's also the question of where should we host them?

> ![Upgrade Data Flow Diagram](imgs/upgrade-dfd.png)

Author: `mermaid` looks incredible - this is a great idea.

> ![Upgrade Data Flow Diagram](imgs/upgrade-dfd.png)

Author: How're these?

https://github.com/awslabs/smithy-rs/pull/1782/commits/e40aff985957469a0de6e171050318384fc24c81

> ![Upgrade Data Flow Diagram](imgs/upgrade-dfd.png)

Reviewer: Looking sharp!

            get_pokemon_species: FailOnMissingOperation,
            get_server_statistics: FailOnMissingOperation,
            capture_pokemon: FailOnMissingOperation,
            get_storage: FailOnMissingOperation,
        }
    }
```

The `builder` constructor provides a `PokemonServiceBuilder` where `build` cannot be called until all operations are set because `MissingOperation` purposefully doesn't implement `Upgradable`. In contrast, the `unchecked_builder` which sets all `Op{N}` to `FailOnMissingOperation` can be immediately built, however any unset operations are upgraded into a service which always returns status code 500, as noted in [Upgrading a Model Service](#upgrading-a-model-service).

After all `Op{N}` are upgraded in `build` they are collected into their protocol specific `Router` implementation and then bundled up into a `RoutingService`. The `RoutingService` is then wrapped in a `PokemonService` newtype and presented to the user.

```rust
/// The Pokémon Service allows you to retrieve information about Pokémon species.
#[derive(Clone)]
pub struct PokemonService<S> {
    router: RoutingService<RestRouter<S>, AwsRestJson1>,
}
```

> ## Plugins

Author: There should be a standalone tutorial on actually writing a `Plugin`. And when there is we should link to it here.

Add a TODO.

All builders implement the `Pluggable` trait, which allows them to apply plugins to service builders:

```rust
/// Provides a standard interface for applying [`Plugin`]s to a service builder. This is implemented automatically for
/// all builders.
///
/// As [`Plugin`]s modify the way in which [`Operation`]s are [`upgraded`](crate::operation::Upgradable) we can use
/// [`Pluggable`] as a foundation to write extension traits which are implemented for all service builders.
///
/// # Example
///
/// ```
/// # struct PrintPlugin;
/// # use aws_smithy_http_server::plugin::Pluggable;
/// trait PrintExt: Pluggable<PrintPlugin> {
///     fn print(self) -> Self::Output where Self: Sized {
///         self.apply(PrintPlugin)
///     }
/// }
///
/// impl<Builder> PrintExt for Builder where Builder: Pluggable<PrintPlugin> {}
/// ```
pub trait Pluggable<NewPlugin> {
    type Output;

    /// Applies a plugin to the service builder.
    fn apply(self, plugin: NewPlugin) -> Self::Output;
}
```

As seen in the `Pluggable` documentation, third-parties can use extension traits over `Pluggable` to extend the API of builders.

> 

Author: I think it's probably a good idea to add the implementation of `Pluggable` on the `PokemonServiceBuilder`.

> ## Accessing Unmodelled Data

Author: This section is quite brief. The full story amounts to pepper the code with `Exts: FromParts<P>` in a predictable way and can be read quite easily from the source code.

Should I go into more detail? Should I include the paragraph above?

```rust
/// An extension trait over [`OperationShape`].
pub trait OperationShapeExt: OperationShape {
    /// Creates a new [`Operation`] for well-formed [`Handler`]s.
    fn from_handler<H>(handler: H) -> Operation<IntoService<Self, H>>
    where
        H: Handler<Self>,
        Self: Sized,
    {
        Operation::from_handler(handler)
    }

    /// Creates a new [`Operation`] for well-formed [`Service`](tower::Service)s.
    fn from_service<S>(svc: S) -> Operation<Normalize<Self, S>>
    where
        S: OperationService<Self>,
        Self: Sized,
    {
        Operation::from_service(svc)
    }
}
```

Observe that there are two constructors provided: `from_handler` which takes a `H: Handler` and `from_service` which takes a `S: OperationService`. In both cases `Self` is passed as a parameter to the traits - this constrains `handler: H` and `svc: S` to the signature given by the implementation of `OperationShape` on `Self`.

> The [`Handler`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/operation/handler.rs#L21-L29) and [`OperationService`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/operation/operation_service.rs#L15-L29) both serve a similar purpose - they provide a common interface for converting to a model service `S`. The `Handler<GetPokemonSpecies>` trait covers all closures taking `GetPokemonSpeciesInput` and asynchronously returning a `Result<GetPokemonSpeciesOutput, GetPokemonSpeciesError>` - they are converted to a model service by a `IntoService`. The `OperationService<GetPokemonSpecies>` trait covers all `tower::Service`s with request `GetPokemonSpeciesInput`, response `GetPokemonSpeciesOutput` and error `GetPokemonSpeciesOutput` - they are converted to a model service by a `Normalize` (this is a very small conversion which flattens request tuples).

Author: There's some detailed documentation in the `operation.rs` module. I'll link to that so they can see the details if they want.

As mentioned, `L` is applied _after_ the `Operation<S, L>` has been "upgraded" to a HTTP service. The procedure of upgrading a model service to a HTTP service is described in the [Upgrading a Model Service](#upgrading-a-model-service) section below.

## Serialization and Deserialization

A [Smithy protocol](https://awslabs.github.io/smithy/2.0/spec/protocol-traits.html#serialization-and-protocol-traits) specifies the serialization/deserialization scheme - how a HTTP request is transformed into a modelled input and a modelled output to a HTTP response. The is formalized using the [`FromRequest`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/request.rs#L156-L164) and [`IntoResponse`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/response.rs#L40-L44) traits:

```rust
/// Provides a protocol aware extraction from a [`Request`]. This consumes the
/// [`Request`], in contrast to [`FromParts`].

> pub trait FromRequest<Protocol>: Sized {

Author: These traits, and others all have a loose counterpart in `axum` e.g. https://docs.rs/axum/latest/axum/extract/trait.FromRequest.html because

1. We used to used `axum` extensively
2. These traits are reasonably natural and occur across all frameworks ([Rocket](https://api.rocket.rs/v0.4/rocket/request/trait.FromRequest.html, [Actix](https://docs.rs/actix-web/latest/actix_web/trait.FromRequest.html), )

Should we mention this?

Reviewer: I don't think it's necessary, it's a fairly common approach in the Rust web ecosystem (as you stated).

{
    type Service = MissingFailure<P>;

    fn upgrade(self, _plugin: &Pl) -> Self::Service {
        MissingFailure { _protocol: PhantomData }
    }
}
```

We go into more detail on how the `Upgradable` trait is used in conjunction with builders in the [Builders](#builders) section below.

## Routers

Different protocols supported by Smithy enjoy different routing mechanisms, for example, [AWS JSON 1.0](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-json-1_0-protocol.html#protocol-behaviors) uses the `X-Amz-Target` header to select an operation, whereas [AWS REST XML](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-restxml-protocol.html) uses the [HTTP label trait](https://awslabs.github.io/smithy/2.0/spec/http-bindings.html#httplabel-trait).

Despite their differences, all routing mechanisms satisfy a common interface. This is formalized using the `Router` trait:

```rust
/// An interface for retrieving an inner [`Service`] given a [`http::Request`].
pub trait Router {
    type Service;
    type Error;

    /// Matches a [`http::Request`] to a target [`Service`].
    fn match_route(&self, request: &http::Request) -> Result<Self::Service, Self::Error>;
}
```

> which provides the ability to lookup an inner HTTP service from a collection using a `&http::Request`.

Reviewer 2: nit: "lookup" gives me an O(1) immediacy sense that is not necessarily true. "determine" might be a better verb?

Types which implement the `Router` trait are converted to a HTTP service via the `RoutingService` struct:

```rust
/// A [`Service`] using a [`Router`] `R` to redirect messages to specific routes.
///
/// The `Protocol` parameter is used to determine the serialization of errors.
pub struct RoutingService<R, Protocol> {
    router: R,
    _protocol: PhantomData<Protocol>,
}

impl<R, P> Service<http::Request> for RoutingService<R, P>
where
    R: Router<B>,
    R::Service: Service<http::Request, Response = http::Response>,
    R::Error: IntoResponse<P> + Error,
{
    type Response = http::Response;

>     type Error = Box<dyn Error>;

Reviewer 2: I'd call out again that we're simplifying the implementation in favor of a more pedagogical text, lest a reader think we're actually boxing every time we return a routing error.

    async fn call(&mut self, req: http::Request<B>) -> Result<Self::Response, Self::Error> {
        match self.router.match_route(&req) {
            // Successfully routed, use the routes `Service::call`.
            Ok(ok) => ok.oneshot(req).await,
            // Failed to route, use the `R::Error`s `IntoResponse<P>`.
            Err(error) => {
                debug!(%error, "failed to route");
                Err(Box::new(error.into_response()))
            }
        }
    }
}
```

The `RouterService` is the final piece necessary to form a functioning composition - it is used to aggregate together the HTTP services, created via the upgrade procedure, into a single HTTP service which can be presented to the customer.

![RouterService](imgs/router.png)

## Builders

> The service builder is the primary public API, it can be generated for every [Smithy Service](https://awslabs.github.io/smithy/2.0/spec/service-types.html). At a high-level, the purpose of a service builder is to take a function for each Smithy Operation, whose signature is constrained by the Smithy model, and produce a single HTTP service.

Reviewer 2: Or closure, or model service, as explained in the beginning.

Author: I'm tempted to leave it as is because a [`Service`](https://docs.rs/tower-service/latest/tower_service/trait.Service.html) is already described as a

>An asynchronous function from a Request to a Response.

And closures are anonymous functions, so this is "true enough".

```rust
/// The service builder for [`PokemonService`].
///
/// Constructed via [`PokemonService::builder`].
pub struct PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6> {
    capture_pokemon_operation: Op1,
    empty_operation: Op2,
    get_pokemon_species: Op3,
    get_server_statistics: Op4,
    get_storage: Op5,
    health_check_operation: Op6,
}
```

> The builder has one generic variable for each [Smithy Operation](https://awslabs.github.io/smithy/2.0/spec/service-types.html#operation) in the [Smithy Service](https://awslabs.github.io/smithy/2.0/spec/service-types.html#service). Each setter switches the type of the `Op{N}` type parameter:

Reviewer 2: ```suggestion
The builder has one generic type parameter for each [Smithy Operation](https://awslabs.github.io/smithy/2.0/spec/service-types.html#operation) in the [Smithy Service](https://awslabs.github.io/smithy/2.0/spec/service-types.html#service). Each setter switches the type of the `Op{N}` type parameter:
```

    /// Sets the [`GetPokemonSpecies`](crate::operation_shape::GetPokemonSpecies) operation.
    ///
    /// This should be an [`Operation`](aws_smithy_http_server::operation::Operation) created from
    /// [`GetPokemonSpecies`](crate::operation_shape::GetPokemonSpecies) using either
    /// [`OperationShape::from_handler`](aws_smithy_http_server::operation::OperationShapeExt::from_handler) or
    /// [`OperationShape::from_service`](aws_smithy_http_server::operation::OperationShapeExt::from_service).
    pub fn get_pokemon_species_operation<NewOp>(
        self,
        value: NewOp,
    ) -> PokemonServiceBuilder<
        Op1,
        Op2,
        NewOp,
        Op4,
        Op5,
        Op6,
    > {
        PokemonServiceBuilder {
            capture_pokemon_operation: self.capture_pokemon_operation,
            empty_operation: self.empty_operation,
            get_pokemon_species: value,
            get_server_statistics: self.get_server_statistics,
            get_storage: self.get_storage,
            health_check_operation: self.health_check_operation,
        }
    }

    /// Sets the [`GetPokemonSpecies`](crate::operation_shape::GetPokemonSpecies) operation.
    ///

>     /// This should be a closure satisfying the [`Handler`](aws_smithy_http_server::operation::Handler) trait.

Reviewer 2: Should we change these docs to highlight this can also be a function? And to mention `async`.

    /// See the [operation module documentation](aws_smithy_http_server::operation) for more information.
    pub fn get_pokemon_species<H, NewExts>(
        self,
        value: H,
    ) -> PokemonServiceBuilder<
        Op1,
        Op2,
        Operation<IntoService<GetPokemonSpecies, H>>,
        Op4,
        Op5,
        Op6,
    >
    where
        H: Handler<GetPokemonSpecies>,
    {
        self.get_pokemon_species_operation(GetPokemonSpecies::from_handler(value))
    }
```

> To finalize the build and construct the complete service, `PokemonService`, each builder has a `build` method whose constraints list out all the requirements for composition:

Reviewer 2: ```suggestion
To finalize the build and construct the complete service, `PokemonService`, each builder has a `build` method whose bounds list out all the requirements for composition:
```

    > {
        PokemonServiceBuilder {
            check_health: MissingOperation,
            do_nothing: MissingOperation,
            get_pokemon_species: MissingOperation,
            get_server_statistics: MissingOperation,
            capture_pokemon: MissingOperation,
            get_storage: MissingOperation,
        }
    }

    /// Constructs an unchecked builder for [`PokemonService`].
    ///
    /// This will not enforce that all operations are set, however if an unset operation is used at runtime
    /// it will return status code 500 and log an error.
    pub fn unchecked_builder() -> PokemonServiceBuilder<
        FailOnMissingOperation,
        FailOnMissingOperation,
        FailOnMissingOperation,
        FailOnMissingOperation,
        FailOnMissingOperation,
        FailOnMissingOperation,
    > {
        PokemonServiceBuilder {
            check_health: FailOnMissingOperation,
            do_nothing: FailOnMissingOperation,
            get_pokemon_species: FailOnMissingOperation,
            get_server_statistics: FailOnMissingOperation,
            capture_pokemon: FailOnMissingOperation,
            get_storage: FailOnMissingOperation,
        }
    }
```

> The `builder` constructor provides a `PokemonServiceBuilder` where `build` cannot be called until all operations are set because `MissingOperation` purposefully doesn't implement `Upgradable`. In contrast, the `unchecked_builder` which sets all `Op{N}` to `FailOnMissingOperation` can be immediately built, however any unset operations are upgraded into a service which always returns status code 500, as noted in [Upgrading a Model Service](#upgrading-a-model-service).

Reviewer 2: I'd omit `unchecked_builder` entirely. It's just going to make the reader raise their eyebrows as to its utility if we don't mention here the multi-Lambda based set up use case, which is not relevant to this text.

Author: I'm on the fence here - if we don't have `unchecked_builder` and `FailOnMissingOperation` then we don't have an example of an alternative `Upgradable::upgrade` path which weakens the motivation for the `Upgradable` trait entirely - which might raise eyebrows.

/// [`Pluggable::apply`].
pub trait Plugin<Protocol, Op, S, L> {
    type Service;
    type Layer;

    /// Maps an [`Operation`] to another.
    fn map(&self, input: Operation<S, L>) -> Operation<Self::Service, Self::Layer>;
}
```

The `Upgradable::upgrade` method on `Operation<S, L>`, previously presented in [Upgrading a Model Service](#upgrading-a-model-service), is more accurately:

```rust
    /// Takes the [`Operation<S, L>`](Operation), applies [`Plugin`], then applies [`UpgradeLayer`] to
    /// the modified `S`, then finally applies the modified `L`.
    ///
    /// The composition is made explicit in the method constraints and return type.
    fn upgrade(self, plugin: &Pl) -> Self::Service {
        let Operation { inner, layer } = plugin.map(self);
        let layer = Stack::new(UpgradeLayer::new(), layer);
        layer.layer(inner)
    }
```

![Upgradable Diagram](imgs/upgradable-plugin.png)

An example `Plugin` implementation can be found in [aws-smithy-http-server/examples/pokemon-service/src/plugin.rs](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/plugin.rs).

> All builders implement the [`Pluggable`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/plugin.rs#L8-L29) trait, which allows them to apply plugins to service builders:

Reviewer 2: ```suggestion
The service builder implements the [`Pluggable`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/plugin.rs#L8-L29) trait, which allows service owners to apply plugins:
```

```rust
/// The service builder for [`PokemonService`].
///
/// Constructed via [`PokemonService::builder`].
pub struct PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6, Pl> {
    capture_pokemon_operation: Op1,
    empty_operation: Op2,
    get_pokemon_species: Op3,
    get_server_statistics: Op4,
    get_storage: Op5,
    health_check_operation: Op6,

    plugin: Pl
}
```

which allows the following `Pluggable` implementation to be generated:

```rust
impl<Op1, Op2, /* ... */, Pl, NewPl> aws_smithy_http_server::plugin::Pluggable<NewPl>
    for PokemonServiceBuilder<Op1, Op2, /* ... */, Pl>
{
    type Output =
        PokemonServiceBuilder<Op1, Exts1, aws_smithy_http_server::plugin::PluginStack<Pl, NewPl>>;
    fn apply(self, plugin: NewPl) -> Self::Output {
        PokemonServiceBuilder {
            capture_pokemon_operation: self.capture_pokemon_operation,
            empty_operation: self.empty_operation,
            /* ... */,

            plugin: PluginStack::new(self.plugin, plugin),
        }
    }
}
```

> Here `PluginStack` works in a similar way to [`tower::layer::util::Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html) - allowing users to append a new plugin rather than replacing the currently set one.

Reviewer: It's not immediately clear to me when a user should opt for a `Plugin` instead of attaching a `Layer` to the overall `Router`. 
A concrete example of common functionality that could only be accomplished with a `Plugin` instead of a `Layer` would ⭐

Reviewer 2: That's a very good observation! Plugins allow us to modify the "upgrade procedure" from model services to HTTP services. There we can "squish in" middlewares that work at the model service layer, whereas the ones applied to the overall Router need to operate at the HTTP service layer, _before_ routing has taken place. You'll notice this difference if you look closely at the diagrams.

You can pull the thread for the primary motivation for introducing plugins in item 2 in the [Service Builder Improvements RFC](https://awslabs.github.io/smithy-rs/design/rfcs/rfc0020_service_builder.html#operations-as-middleware-constructors), to perform operation-specific logging behavior, but there are other use cases where we want to modify behavior at the model service layer.

Author: How's this?
https://github.com/awslabs/smithy-rs/pull/1782/commits/8ff688e219a6dd41fc73cf65646116d0e657f52f

There's an unfortunate conflict, [which exists for now](https://github.com/awslabs/smithy-rs/issues/1707), between the [`Router`](https://github.com/awslabs/smithy-rs/blob/89a6e5894aebb81efa1305c41188d7ded3b3c43b/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L64-L67) struct, which is essentially an enum of the [protocol specific routers](https://github.com/awslabs/smithy-rs/blob/89a6e5894aebb81efa1305c41188d7ded3b3c43b/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L69-L82) and the [`Router`](https://github.com/awslabs/smithy-rs/blob/89a6e5894aebb81efa1305c41188d7ded3b3c43b/rust-runtime/aws-smithy-http-server/src/routers/mod.rs#L40-L47) trait which is an interface for routing requests to services. This makes things more confusing for people today.

> Here `PluginStack` works in a similar way to [`tower::layer::util::Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html) - allowing users to append a new plugin rather than replacing the currently set one.

Reviewer 2: ```suggestion
Here `PluginStack` works in a similar way to [`tower::layer::util::Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html) - allowing users to compose plugins rather than replacing the currently set one.
```

compose/stack/layer might be better verbs than append.

This survey is disinterested in the actual Kotlin implementation of the code generator, and instead focuses on the structure of the generated Rust code and how it relates to the Smithy model. The intended audience is new contributors and users interested in internal details.

During the survey we will use the [`pokemon.smithy`](https://github.com/awslabs/smithy-rs/blob/main/codegen-core/common-test-models/pokemon.smithy) model as a reference:

```smithy
/// The Pokémon Service allows you to retrieve information about Pokémon species.
@title("Pokémon Service")
@restJson1
service PokemonService {
    version: "2021-12-01",
    resources: [PokemonSpecies, Storage],
    operations: [
        GetServerStatistics,
        DoNothing,
        CapturePokemon,

>         CheckHealth

Reviewer 3: The `GetPokemonSpecies` operation is not listed here, but is used throughout the examples, that's a bit confusing.

We represent this in Rust using the [`OperationShape`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/operation/shape.rs#L8-L22) trait:

```rust
pub trait OperationShape {
    /// The name of the operation.
    const NAME: &'static str;

    /// The operation input.
    type Input;
    /// The operation output.
    type Output;
    /// The operation error. [`Infallible`](std::convert::Infallible) in the case where no error
    /// exists.
    type Error;
}
```

For each Smithy Operation shape,

```smithy
/// Retrieve information about a Pokémon species.
@readonly
@http(uri: "/pokemon-species/{name}", method: "GET")
operation GetPokemonSpecies {
    input: GetPokemonSpeciesInput,
    output: GetPokemonSpeciesOutput,
    errors: [ResourceNotFoundException],
}
```

the following implementation is generated

```rust
/// Retrieve information about a Pokémon species.
pub struct GetPokemonSpecies;

impl OperationShape for GetPokemonSpecies {
    const NAME: &'static str = "com.aws.example#GetPokemonSpecies";

    type Input = GetPokemonSpeciesInput;
    type Output = GetPokemonSpeciesOutput;

>     type Error = GetPokemonSpeciesError;

Reviewer 3: This type, `GetPokemonSpeciesError`, doesn't appear in the smithy file, I assume it's a generated (rust) enum of the possible errors, but maybe it could use an introduction?

- For a specific operation, for example `GetPokemonSpecies`, the model service can be wrapped by a `Layer` before passing it to `GetPokemonSpecies::from_service` constructor.
- The `Operation::layer` method can be used to apply a `Layer` to a specific operation _after_ it's been upgraded.

After the build is finalized:

- The entire `PokemonService` HTTP service can be wrapped by a `Layer`.
- Every `Route` in the `Router` can be wrapped by a `Layer` using `PokemonService::layer`.

Although this provides a reasonably "complete" API, it can be cumbersome in some use cases. Suppose a customer wants to log the operation name when a request is routed to said operation. Writing a `Layer`, `NameLogger`, to log a operation name is simple, however with the current API the customer is forced to do the following

```rust
let get_pokemon_species = GetPokemonSpecies::from_handler(/* handler */).layer(NameLogger::new("GetPokemonSpecies"));
let get_storage = GetStorage::from_handler(/* handler */).layer(NameLogger::new("GetStorage"));
let do_nothing = DoNothing::from_handler(/* handler */).layer(NameLogger::new("DoNothing"));
/* Repeat for every route... */
```

> Note that `PokemonService::layer` cannot be used here because it applies a _single_ layer uniformly across all `Route`s stored in the `Router`.

Reviewer: If I understood the machinery correctly, this is not necessarily the problem per se.
Our problem is that the layer we are applying does not access to the `Operation` struct, therefore it has no way to ergonomically retrieve the data that we want in this example (i.e. the name of the operation).

Author: Yeah, by the time `PokemonService` exists we've already wrapped them in a `Route` and collected them into some `Router` implementation.

We currently don't hold any additional information other than the model service `S` and the layer `L` in the `Operation`, so if the `Layer` did have access to the specific `Operation` that had been bundled into the `Route` it still wouldn't be able to make any judgements about the name. Furthermore, suppose a `FailOnMissing` had been the source of the `Route`, there's no operation name for it.
